### PR TITLE
Adding new argument `max_new_tokens` for generate.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

This is a proposal to add a new argument `max_new_tokens` to `generate`.
This include a `MaxNewTokensCriteria` that enables callers that don't
know about the token length ahead (like pipelines callers) to manage
more easily the length of their generated output.

`max_length` is a hard to use argument for generate:

- It means different things in `encoder-decoder` context and
`decoder-only` context
  - `encoder-decoder`: max_length = max_new_tokens - 1 (in case of bos)
  - `decoder-only`: max_length = input_ids.shape[-1] + max_new_tokens
- It is hard to understand from a pipeline point of view where `tokens`
do not exist yet.

This is a proposal to add a new argument `max_new_tokens` to `generate`.
This include a `MaxNewTokensCriteria` which is a bit redundant with
respect to `MaxLengthCriteria`. It is a consistency concern for now but debattable.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@LysandreJik  @patrickvonplaten 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Narsil** [comment]
@LysandreJik @patrickvonplaten

Forgot to add you for review, my bad.

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -690,8 +692,11 @@ def generate(
             input_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`, `optional`):
                 The sequence used as a prompt for the generation. If :obj:`None` the method initializes it as an empty
                 :obj:`torch.LongTensor` of shape :obj:`(1,)`.
-            max_length (:obj:`int`, `optional`, defaults to 20):
+            max_length (:obj:`int`, `optional`, defaults to :obj:`model.config.max_length`):
                 The maximum length of the sequence to be generated.
+            max_new_tokens (:obj:`int`, `optional`, defaults to None):
```
Comment (LysandreJik) [inline src/transformers/generation_utils.py:697]: No need to specify `defaults to None` for `optional` values 
```suggestion
            max_new_tokens (:obj:`int`, `optional`):
```

**Review by LysandreJik — APPROVED** [review]
I like the proposal and I know this would have been clearer to me in the past as well - will defer to @patrickvonplaten regarding the actual implementation.

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -690,8 +692,11 @@ def generate(
             input_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`, `optional`):
                 The sequence used as a prompt for the generation. If :obj:`None` the method initializes it as an empty
                 :obj:`torch.LongTensor` of shape :obj:`(1,)`.
-            max_length (:obj:`int`, `optional`, defaults to 20):
+            max_length (:obj:`int`, `optional`, defaults to :obj:`model.config.max_length`):
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:695]: That's actually a good idea! We should probably refactor most of the docstrings like this

File: src/transformers/generation_utils.py [hunk 3]
```diff
@@ -958,13 +972,10 @@ def generate(
             remove_invalid_values=remove_invalid_values,
         )
 
-        stopping_criteria = self._get_stopping_criteria(max_length=max_length, max_time=max_time)
-        if max_length is not None:
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:962]: This shouldn't have been there before neither no?

File: src/transformers/generation_utils.py [hunk 4]
```diff
@@ -690,8 +692,11 @@ def generate(
             input_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`, `optional`):
                 The sequence used as a prompt for the generation. If :obj:`None` the method initializes it as an empty
                 :obj:`torch.LongTensor` of shape :obj:`(1,)`.
-            max_length (:obj:`int`, `optional`, defaults to 20):
+            max_length (:obj:`int`, `optional`, defaults to :obj:`model.config.max_length`):
                 The maximum length of the sequence to be generated.
+            max_new_tokens (:obj:`int`, `optional`):
+                The maximum numbers of tokens to generate, ignore the current number of tokens. Use either
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:698]:
```suggestion
                The maximum numbers of tokens to generate, independent of the input length. It is recommended to either use :obj:`max_new_tokens` or :obj:`max_length`, but not both.
```

File: tests/test_generation_utils.py [hunk 5]
```diff
@@ -1577,3 +1577,29 @@ def test_beam_search_warning_if_max_length_is_passed(self):
...
+
+        # Encoder decoder call
+        max_new_tokens = 3
+        with torch.no_grad():
+            outputs = bart_model.generate(input_ids, max_new_tokens=max_new_tokens)
+        # 1 BOS + 3 new tokens
+        self.assertEqual(list(outputs.shape), [1, 4])
+
+        # Decoder only call
+        with torch.no_grad():
```
Comment (patrickvonplaten) [inline tests/test_generation_utils.py:1597]: `generate` is always wrapped in `torch.no_grad()` no need for this here

File: tests/test_generation_utils.py [hunk 6]
```diff
@@ -1577,3 +1577,29 @@ def test_beam_search_warning_if_max_length_is_passed(self):
...
+            outputs = bart_model.generate(decoder_input_ids=input_ids, max_new_tokens=max_new_tokens)
+        # 15 + 3 new tokens
+        self.assertEqual(list(outputs.shape), [1, 18])
+
+        with torch.no_grad():
```
Comment (patrickvonplaten) [inline tests/test_generation_utils.py:1602]:
```suggestion
```

**Review by patrickvonplaten — APPROVED** [review]
Left some nits

File: src/transformers/generation_utils.py (hunk 3 above)
Comment (Narsil) [reply src/transformers/generation_utils.py:962]: No it was fixed in another PR, it should be removed after rebase.

**Pull request merged**